### PR TITLE
Fix core update when 2FA or MFA is enabled

### PR DESCRIPTION
**Pull request by nikosdion**
Pull Request for Issue #38044 .

### Summary of Changes

Allows Joomla! Update to finalise if you had previously enabled 2FA or MFA for the Super User executing the update.

### Testing Instructions

Update a Joomla 4.1.x stable or development version or a 4.2 development version prior to 4.2.0-beta1 with Two Factor Authentication enabled and used for the current super user who runs the update to a 4.2.0-beta2 or current 4.2-dev version which includes the new Multi Factor Authentication introduced with PR #37912 .

### Actual result BEFORE applying this Pull Request

When the update package has been unpacked (i.e. progress bar comes to the end), the update ends with an SQL error about the new table for MFA missing.

### Expected result AFTER applying this Pull Request

The update completes cleanly.

### Documentation Changes Required

None.

### CODE REVIEW

**Comment by nikosdion** [comment]
@richard67 Please mark as release blocker

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 67a902d295ff56038301d8fe3a3e75978a20a045<br /><br />Hint for other testers: To reproduce the issue, I've updated with 2FA enabled and used to the custom update URL of the 4.2-dev nightlies.

To test the fix, I've updated to the custom update URL created by drone for this PR and which can currently be found here: ~~https://ci.joomla.org/artifacts/joomla/joomla-cms/4.2-dev/38045/downloads/55243/~~
Update: New link is https://ci.joomla.org/artifacts/joomla/joomla-cms/4.2-dev/38045/downloads/55281/ .

With the fix, you get prompted for your 2FA (or now MFA) key or password or whatever it is (in my case the 6 digit code for Google Authenticator) on the new captive MFA page. After having submitted that, you see the normal success message for the update, and all is fine.



<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38045">issues.joomla.org/tracker/joomla-cms/38045</a>.</sub>

**Comment by richard67** [comment]
Screenshots of my test result:
2022-06-13_test-pr-38045_1

After entering the code and validate:
2022-06-13_test-pr-38045_2

**Comment by richard67** [comment]
P.S.: Log out and log in again works as it should with the new MFA.

**Comment by nikosdion** [comment]
That's a perfect test! Exactly how it should've worked in the first place 😊

**Comment by richard67** [comment]
Yes, and all is like it should be in the MFA options. The code is there, backup keys are there ... all fine ... and I like the new MFA.

**Comment by nikosdion** [comment]
Thank you :) The new MFA is what I've always had in mind ever since I proposed integrating 2FA back in 2013. Finally the time was right to get this into the core. Onwards and upwards!

File: libraries/src/Application/MultiFactorAuthenticationHandler.php [hunk 1]
```diff
@@ -311,6 +311,12 @@ private function needsMultiFactorAuthenticationRedirection(): bool
 			return false;
 		}
 
+		// Allow the Joomla update finalisation to run
+		if ($isAdmin && $option === 'com_joomlaupdate' && in_array($task, ['update.finalise', 'update.cleanup']))
```
Comment (zero-24) [inline libraries/src/Application/MultiFactorAuthenticationHandler.php]:
```suggestion
		if ($isAdmin && $option === 'com_joomlaupdate' && in_array($task, ['update.finalise', 'update.cleanup', 'update.finaliseconfirm']))
```

Isnt [finalisecofirm ](https://github.com/joomla/joomla-cms/blob/4.2-dev/administrator/components/com_joomlaupdate/src/Controller/UpdateController.php#L433) also to be allowed in some cases? Or I'm wrong here?

Comment (nikosdion) [reply libraries/src/Application/MultiFactorAuthenticationHandler.php]: Good catch

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 7de2f9a59483f068aa9a58b94338269445bb810a<br /><br />This time I've tested not only live update but also upload & update to be sure I haven't forgotten anything. Both works fine.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38045">issues.joomla.org/tracker/joomla-cms/38045</a>.</sub>

**Comment by N6REJ** [comment]
I was able to take a live 4.1.4 site, use the nightly build for 4.2 both with and without 2fa and it updated successfully.

**Comment by nikosdion** [comment]
@N6REJ Can you please issues.joomla.org to submit a successful test? This is the only way for GitHub to set the JTracker/HumanTestResults check as successful.

**Comment by HLeithner** [comment]
Tested successful thanks

**Pull request merged**